opm/autodiff/StandardWellsDense_impl.hpp

```diff
@@ -1987,21 +1989,29 @@ namespace Opm {
 
             // when it is a single phase rate limit
             if (num_phases_under_rate_control == 1) {
-                if (distr[phaseIdx] == 1.0) {
-                    qs.setValue(target_rate);
-                    return qs;
+
+                // looking for the phase under control
+                int phase_under_control = -1;
+                for (int phase = 0; phase < np; ++phase) {
+                    if (distr[phase] > 0.0) {
+                        phase_under_control = phase;
+                        break;
+                    }
                 }
 
-                int currentControlIdx = 0;
-                for (int i = 0; i < np; ++i) {
-                    currentControlIdx += wells().comp_frac[np*wellIdx + i] * i;
+                assert(phase_under_control >= 0);
+
+                if (phaseIdx == phase_under_control) {
+                    qs.setValue(target_rate);
+                    return qs;
                 }
 
+                // TODO: not sure why the single phase under control will have near zero fraction
                 const double eps = 1e-6;
```

Comment:
If you change from gas to water injection. Your well may be completely filled with gas initially giving a zero or near zero fraction. After one well iteration this should no longer be a problem. Maybe zero rated well could also trigger problems here.